File: src/utils/client-script.js
#!/usr/bin/env node
// DDNS Client Script
// This script automatically updates your dynamic DNS record
// Token and server URL will be replaced with actual values when served

const TOKEN = '__TOKEN__';
const SERVER_URL = '__SERVER_URL__';
const UPDATE_INTERVAL = 300000; // 5 minutes
const DOMAIN = '__DOMAIN__';

// Use built-in Node.js modules to avoid dependencies
const http = require('http');
const https = require('https');
const { URL } = require('url');

/**
 * Make an HTTP request using native Node.js modules
 * @param {string} url - The URL to request
 * @param {object} options - Request options
 * @param {object|string} [data] - Request body data for POST/PUT
 * @returns {Promise<object>} - Promise resolving to the response
 */
function makeRequest(url, options = {}, data = null) {
  return new Promise((resolve, reject) => {
    // Parse the URL to determine protocol (http or https)
    const parsedUrl = new URL(url);
    const client = parsedUrl.protocol === 'https:' ? https : http;
    
    // Prepare request options
    const requestOptions = {
      method: options.method || 'GET',
      headers: options.headers || {},
      ...options
    };
    
    // Create request
    const req = client.request(url, requestOptions, (res) => {
      // Handle redirects (follow up to 5 redirects)
      if (res.statusCode >= 300 && res.statusCode < 400 && res.headers.location) {
        if (options._redirectCount && options._redirectCount >= 5) {
          return reject(new Error('Too many redirects'));
        }
        
        // Update redirect count
        const redirectOptions = {
          ...options,
          _redirectCount: (options._redirectCount || 0) + 1
        };
        
        // Follow redirect
        return resolve(makeRequest(res.headers.location, redirectOptions, data));
      }
      
      // Collect response data
      let responseData = '';
      res.setEncoding('utf8');
      
      res.on('data', (chunk) => {
        responseData += chunk;
      });
      
      res.on('end', () => {
        try {
          // Try to parse as JSON
          const parsedData = responseData.trim() ? JSON.parse(responseData) : {};
          
          if (res.statusCode >= 200 && res.statusCode < 300) {
            resolve(parsedData);
          } else {
            const error = new Error(`Request failed with status ${res.statusCode}`);
            error.response = {
              status: res.statusCode,
              data: parsedData,
              headers: res.headers
            };
            reject(error);
          }
        } catch (e) {
          // Handle non-JSON responses
          if (res.statusCode >= 200 && res.statusCode < 300) {
            resolve({ data: responseData });
          } else {
            const error = new Error(`Request failed with status ${res.statusCode}`);
            error.response = {
              status: res.statusCode,
              data: responseData,
              headers: res.headers
            };
            reject(error);
          }
        }
      });
    });
    
    // Handle request errors
    req.on('error', (err) => {
      reject(err);
    });
    
    // Send data if provided (for POST/PUT)
    if (data) {
      const bodyData = typeof data === 'string' ? data : JSON.stringify(data);
      req.write(bodyData);
    }
    
    // End the request
    req.end();
  });
}

/**
 * Format timestamp for logging
 * @returns {string} Formatted timestamp [HH:MM:SS]
 */
function timestamp() {
  const now = new Date();
  return `[${now.toISOString().substring(11, 19)}]`;
}

/**
 * Log with timestamp and prefix
 * @param {string} level - Log level (INFO, ERROR, etc.)
 * @param {string} message - Message to log
 */
function log(level, message) {
  console.log(`${timestamp()} ${level}: ${message}`);
}

/**
 * Log error with details
 * @param {string} context - Error context
 * @param {Error} error - Error object
 */
function logError(context, error) {
  log('ERROR', `${context}: ${error.message}`);
  if (error.response) {
    log('ERROR', `Server response: ${JSON.stringify(error.response.data)}`);
  }
  if (error.stack && process.env.DEBUG) {
    console.error(error.stack);
  }
}

// Get current public IP address
async function getCurrentIP() {
  try {
    log('INFO', 'Detecting public IP address...');
    // Use the server's built-in IP detection endpoint
    const response = await makeRequest(`${SERVER_URL}/api/ip`);
    log('INFO', `Successfully detected IP: ${response.ip}`);
    return response.ip;
  } catch (error) {
    logError('Failed to detect IP address', error);
    throw new Error('Failed to detect IP address');
  }
}

// Update DNS record
async function updateDNS(ip) {
  try {
    log('INFO', `Updating DNS record for ${DOMAIN} to ${ip}...`);
    const response = await makeRequest(`${SERVER_URL}/api/dns/update`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${TOKEN}`,
        'Content-Type': 'application/json'
      }
    }, { ip });

    if (response.success) {
      if (response.result && response.result.status === 'unchanged') {
        log('INFO', `DNS record is already up-to-date (${ip})`);
      } else {
        log('INFO', `DNS record updated successfully to ${ip}`);
      }
    }

    return response;
  } catch (error) {
    logError('Failed to update DNS record', error);
    throw error;
  }
}

// Main function to check and update IP
async function checkAndUpdateIP() {
  try {
    const currentIP = await getCurrentIP();
    const result = await updateDNS(currentIP);

    // Schedule next check
    const nextCheckTime = new Date(Date.now() + UPDATE_INTERVAL);
    const nextCheckTimeStr = nextCheckTime.toLocaleTimeString();
    log('INFO', `Next check scheduled at ${nextCheckTimeStr}`);
  } catch (error) {
    log('ERROR', `Update cycle failed: ${error.message}`);
    log('INFO', `Will retry in ${UPDATE_INTERVAL/60000} minutes`);
  }
}

// Run initially and then at intervals
log('INFO', '========================================');
log('INFO', `DDNS client starting for domain: ${DOMAIN}`);
log('INFO', `Server URL: ${SERVER_URL}`);
log('INFO', `Update interval: ${UPDATE_INTERVAL/60000} minutes`);
log('INFO', '========================================');

checkAndUpdateIP();
setInterval(checkAndUpdateIP, UPDATE_INTERVAL);

// Handle termination
process.on('SIGINT', () => {
  log('INFO', 'Received termination signal. Shutting down...');
  log('INFO', 'DDNS client stopped');
  process.exit(0);
});

// Handle unexpected errors
process.on('uncaughtException', (error) => {
  logError('Uncaught exception', error);
  log('ERROR', 'An unexpected error occurred. Client will attempt to continue...');
});

File: src/utils/client-script.py
#!/usr/bin/env python3
# DDNS Client Script (Python Implementation)
# This script automatically updates your dynamic DNS record
# Token and server URL will be replaced with actual values when served

import json
import time
import signal
import sys
from datetime import datetime, timedelta
from urllib.request import Request, urlopen
from urllib.error import URLError, HTTPError
from urllib.parse import urlparse

TOKEN = '__TOKEN__'
SERVER_URL = '__SERVER_URL__'
UPDATE_INTERVAL = 300  # 5 minutes (in seconds)
DOMAIN = '__DOMAIN__'

def timestamp():
    """Format timestamp for logging [HH:MM:SS]"""
    return f"[{datetime.now().strftime('%H:%M:%S')}]"

def log(level, message):
    """Log with timestamp and prefix"""
    print(f"{timestamp()} {level}: {message}")

def log_error(context, error):
    """Log error with details"""
    log("ERROR", f"{context}: {str(error)}")
    if hasattr(error, 'response') and error.response:
        log("ERROR", f"Server response: {error.response}")
    
    import os
    if hasattr(error, 'traceback') and os.environ.get('DEBUG'):
        log("ERROR", error.traceback)

def make_request(url, method="GET", headers=None, data=None, max_redirects=5):
    """
    Make an HTTP request using Python's urllib
    
    Args:
        url: The URL to request
        method: Request method (GET, POST, etc.)
        headers: Request headers dictionary
        data: Request body data for POST/PUT
        max_redirects: Maximum number of redirects to follow
    
    Returns:
        Parsed JSON response or text response
    """
    redirect_count = 0
    
    while True:
        if headers is None:
            headers = {}
            
        req = Request(url, headers=headers, method=method)
        
        if data:
            if isinstance(data, dict):
                data = json.dumps(data).encode('utf-8')
                if 'Content-Type' not in headers:
                    req.add_header('Content-Type', 'application/json')
            elif isinstance(data, str):
                data = data.encode('utf-8')
                
            req.data = data
        
        try:
            response = urlopen(req)
            
            # Handle redirects
            if 300 <= response.status < 400 and 'Location' in response.headers:
                if redirect_count >= max_redirects:
                    raise Exception("Too many redirects")
                
                redirect_count += 1
                url = response.headers['Location']
                continue
            
            # Get response data
            response_data = response.read().decode('utf-8')
            
            # Try to parse JSON
            if response_data.strip():
                try:
                    return json.loads(response_data)
                except json.JSONDecodeError:
                    return {"data": response_data}
            else:
                return {}
                
        except HTTPError as e:
            # Handle HTTP errors
            error_message = f"Request failed with status {e.code}"
            
            try:
                response_data = e.read().decode('utf-8')
                if response_data.strip():
                    try:
                        error_response = json.loads(response_data)
                    except json.JSONDecodeError:
                        error_response = {"data": response_data}
                else:
                    error_response = {}
            except:
                error_response = {}
                
            e.response = {
                "status": e.code,
                "data": error_response,
                "headers": dict(e.headers)
            }
            
            raise Exception(error_message) from e
            
        except URLError as e:
            raise Exception(f"Request failed: {str(e)}") from e
        
        break

async def get_current_ip():
    """Get current public IP address"""
    try:
        log("INFO", "Detecting public IP address...")
        # Use the server's built-in IP detection endpoint
        response = make_request(f"{SERVER_URL}/api/ip")
        log("INFO", f"Successfully detected IP: {response['ip']}")
        return response["ip"]
    except Exception as error:
        log_error("Failed to detect IP address", error)
        raise Exception("Failed to detect IP address")

async def update_dns(ip):
    """Update DNS record"""
    try:
        log("INFO", f"Updating DNS record for {DOMAIN} to {ip}...")
        headers = {
            "Authorization": f"Bearer {TOKEN}",
            "Content-Type": "application/json"
        }
        response = make_request(
            f"{SERVER_URL}/api/dns/update",
            method="POST",
            headers=headers,
            data={"ip": ip}
        )

        if response.get("success"):
            result = response.get("result", {})
            if result and result.get("status") == "unchanged":
                log("INFO", f"DNS record is already up-to-date ({ip})")
            else:
                log("INFO", f"DNS record updated successfully to {ip}")

        return response
    except Exception as error:
        log_error("Failed to update DNS record", error)
        raise error

async def check_and_update_ip():
    """Main function to check and update IP"""
    try:
        current_ip = await get_current_ip()
        result = await update_dns(current_ip)

        # Schedule next check
        next_check_time = (datetime.now() + timedelta(seconds=UPDATE_INTERVAL)).strftime("%H:%M:%S")
        log("INFO", f"Next check scheduled at {next_check_time}")
    except Exception as error:
        log("ERROR", f"Update cycle failed: {str(error)}")
        log("INFO", f"Will retry in {UPDATE_INTERVAL/60} minutes")

def handle_signal(signum, frame):
    """Handle termination signal"""
    log("INFO", "Received termination signal. Shutting down...")
    log("INFO", "DDNS client stopped")
    sys.exit(0)

# Register signal handler
signal.signal(signal.SIGINT, handle_signal)
signal.signal(signal.SIGTERM, handle_signal)

if __name__ == "__main__":
    # Print startup information
    log("INFO", "========================================")
    log("INFO", f"DDNS client starting for domain: {DOMAIN}")
    log("INFO", f"Server URL: {SERVER_URL}")
    log("INFO", f"Update interval: {UPDATE_INTERVAL/60} minutes")
    log("INFO", "========================================")

    # Main loop
    try:
        while True:
            try:
                # In Python, we need to use a simple event loop since asyncio
                # isn't as straightforward for this script's purpose
                import asyncio
                asyncio.run(check_and_update_ip())
                time.sleep(UPDATE_INTERVAL)
            except Exception as e:
                log_error("Uncaught exception", e)
                log("ERROR", "An unexpected error occurred. Client will attempt to continue...")
                time.sleep(UPDATE_INTERVAL)
    except KeyboardInterrupt:
        handle_signal(signal.SIGINT, None)